File: src/game.cpp
// File:    game.cpp
// Purpose: Game class
// Author:  Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include "board.hpp"
#include "game.hpp"
#include "partial.hpp"
#include "strings.hpp"


// lifecycle

Game::Game(
    Strings const &rPlayerNames,
	Indices const &rAutoFlags,
    GameStyleType style,
    unsigned tileRedundancy,
    unsigned handSize,
	unsigned secondsPerHand)
{
	ASSERT(!rPlayerNames.IsEmpty());
	ASSERT(style != GAME_STYLE_NONE);
	ASSERT(tileRedundancy >= 1);
	ASSERT(handSize >= 1);

    // save parameters
	mStyle = style;
	mRedundancy = tileRedundancy;
	mHandSize = handSize;
	mSecondsPerHand = secondsPerHand;

	// add tiles to the stock bag
	for (unsigned i = 0; i < mRedundancy; i++) {
        // generate all possible tiles
		mStockBag.Restock();
	}
    std::cout << "Placed " << plural(CountStock(), "tile") << " in the stock bag." << std::endl;
    
    // construct hands and generate a unique name for each one
    Strings unique = rPlayerNames.Unique();
	Strings::ConstIterator i_player;
	unsigned i = 0;
	for (i_player = rPlayerNames.Begin(); i_player != rPlayerNames.End(); i_player++) {
		String player_name = *i_player;

		String hand_name = player_name;
		if (rPlayerNames.Count(player_name) > 1) {
		    hand_name = unique.InventUnique(player_name, "'s ", " hand");
		    unique.Append(hand_name);
		}

		Hand hand(hand_name, player_name, rAutoFlags.Contains(i));
	    mHands.push_back(hand);
		i++;
    }

    // deal tiles to each hand from the stock bag
    Hands::Iterator i_hand;
    for (i_hand = mHands.begin(); i_hand < mHands.end(); i_hand++) {
        Tiles tiles = i_hand->DrawTiles(mHandSize, mStockBag);

		// record the deal in mHistory
		String name = i_hand->Name();
		Turn turn(tiles, name);
		mHistory.push_back(turn);
    }
    std::cout << std::endl;

    // the hand with the best "run" gets the first turn
	FindBestRun();

	// set iterator to end of history
    miRedo = mHistory.end();

	// pretend this game has been saved, for convenience
	mUnsavedChanges = false;

	ASSERT(IsPaused());
	ASSERT(!CanUndo());
	ASSERT(!CanRedo());
}


// operators

Game::operator Board(void) const {
    return mBoard;
}

Game::operator Hand(void) const {
    Hand result = *miActiveHand;
    
    return result;
}

Game::operator Hands(void) const {
	return mHands;
}


// misc methods

void Game::ActivateNextHand(void) {
	ASSERT(!IsClockRunning());

    // skip over hands which have resigned
    mHands.NextWorking(miActiveHand);
	mUnsavedChanges = true;

	ASSERT(!IsClockRunning());
}

// get copies of the tiles in the active hand
Tiles Game::ActiveTiles(void) const {
    Tiles result = Tiles(*miActiveHand);
    
    return result;
}

void Game::AddTurn(Turn const &rTurn) {
    while (miRedo != mHistory.end()) {
        miRedo = mHistory.erase(miRedo);
    }
    mHistory.push_back(rTurn);
    miRedo = mHistory.end();

	mUnsavedChanges = true;
}

unsigned Game::CountStock(void) const {
    unsigned result = mStockBag.Count();
    
    return result;
}

void Game::DisplayScores(void) const {
    Hands::ConstIterator i_hand;

	std::cout << std::endl;
    for (i_hand = mHands.begin(); i_hand < mHands.end(); i_hand++) {
	    i_hand->DisplayScore();
    }
}

void Game::DisplayStatus(void) const {
    DisplayScores();
    unsigned stock = CountStock();
    std::cout << std::endl
        << miActiveHand->Name() << "'s turn, " 
		<< plural(stock, "tile") << " remaining in the stock bag"
        << std::endl << std::endl << String(mBoard) << std::endl;
}

void Game::FindBestRun(void) {
    Hands::Iterator i_hand;
    mBestRunLength = 0;
    for (i_hand = mHands.begin(); i_hand < mHands.end(); i_hand++) {
		String hand_name = i_hand->Name();
        Tiles run = i_hand->LongestRun();
	    unsigned run_length = run.Count();
        std::cout << hand_name << " has a run of " << plural(run_length, "tile") 
			      << "." << std::endl;

	    if (run_length > mBestRunLength) {
	   	    mBestRunLength = run_length;
		    miActiveHand = i_hand;
	    }
    }
}

void Game::FinishTurn(Move const &rMove) {
    ASSERT(mBoard.IsValidMove(rMove));
    ASSERT(IsClockRunning());
    
    StopClock();
    
    String hand_name = miActiveHand->Name();
    Turn turn(rMove, hand_name, mBestRunLength);

	if (rMove.IsResign()) {
		miActiveHand->Resign(mStockBag);

	} else {
        // remove played/swapped tiles from the hand
        Tiles tiles = Tiles(rMove);
        miActiveHand->RemoveTiles(tiles);

        // attempt to draw replacement tiles from the stock bag
        unsigned count = tiles.Count();
	    if (count > 0) {
            Tiles draw = miActiveHand->DrawTiles(count, mStockBag);
            ASSERT(draw.Count() == count || !rMove.InvolvesSwap());
		    turn.SetDraw(draw);
	    }

	    if (rMove.InvolvesSwap()) {
		    // return swapped tiles to the stock bag
		    mStockBag.AddTiles(tiles);
		    std::cout << miActiveHand->Name() << " put " << plural(count, "tile")
			    << " back into the stock bag." << std::endl;

	    } else {
	        // place played tiles on the board
            mBoard.PlayMove(rMove);
    
            // update the hand's score    
  	        unsigned points = mBoard.ScoreMove(rMove);
	        miActiveHand->AddScore(points);
	        turn.SetPoints(points);
		}
	}
	
	AddTurn(turn);
	
	//  If it was the first turn, it no longer is.
    mBestRunLength = 0;

	// There are now unsaved changes.
	mUnsavedChanges = true;
    ASSERT(!IsClockRunning());
}

void Game::FirstTurn(void) {
    ASSERT(IsPaused());
    std::cout << std::endl;

    Move move;
    StartClock();
    for (;;) {
  	    std::cout << miActiveHand->Name() << " plays first and must place " 
			<< plural(mBestRunLength, "tile") << " on the (empty) board." 
            << std::endl;
        if (miActiveHand->IsAutomatic()) {
			Tiles run = miActiveHand->LongestRun();
            move = Move(run);
		    std::cout << miActiveHand->Name() << " played " << String(move) << std::endl;
			break;
        } else {
		    move = miActiveHand->ChooseMove();
			char const *reason;
    	    if (IsLegalMove(move, reason)) {
                break;
            }
			String title;
			String message = Board::ReasonMessage(reason, title);
			std::cout << message << std::endl << std::endl;
        }
	}
    ASSERT(IsLegalMove(move));

    ASSERT(!IsPaused());
    FinishTurn(move);
    ASSERT(!IsClockRunning());
}

void Game::GoingOutBonus(void) {
    ASSERT(IsOver());
    
	// the active hand scores a point for each tile in every other hand
    Hands::ConstIterator i_hand = miActiveHand;
	mHands.Next(i_hand);
    
    while (i_hand != miActiveHand) {
        Tiles hand = Tiles(*i_hand);
        unsigned pointsInHand = hand.Count(); // TODO
        miActiveHand->AddScore(pointsInHand);
		mHands.Next(i_hand);
    }
	mUnsavedChanges = true;
}

unsigned Game::HandSize(void) const {
	ASSERT(mHandSize > 0);

	return mHandSize;
}

Hands Game::InactiveHands(void) const {
	// return a list of inactive hands (including those which have resigned)
    Hands result;

    Hands::ConstIterator i_hand = miActiveHand;
	mHands.Next(i_hand);
    
    while (i_hand != miActiveHand) {
        result.push_back(*i_hand);
        mHands.Next(i_hand);
    }
    
    return result;
}

void Game::NextTurn(void) {
    ASSERT(IsPaused());

	StartClock();
    DisplayStatus();

    Move move;
    if (miActiveHand->IsAutomatic()) {
		ASSERT(!CanRedo());
        Partial partial(this, HINT_NONE);
        partial.Suggest();
        move = partial.GetMove(false);
		std::cout << miActiveHand->Name() << " played " << String(move) << std::endl;

	} else {
		ASSERT(miActiveHand->IsLocalPlayer());
	    for (;;) {
		    move = miActiveHand->ChooseMove();
			char const *reason;
    	    if (IsLegalMove(move, reason)) {
                break;
            }
			String title;
			String message = Board::ReasonMessage(reason, title);
			std::cout << message << std::endl << std::endl;
			DisplayStatus();
        }
	}
    ASSERT(IsLegalMove(move));

    ASSERT(!IsPaused());
    FinishTurn(move);
    ASSERT(!IsClockRunning());
}

void Game::PlayGame(void) {
    FirstTurn();

    while (!IsOver()) {
        ActivateNextHand();
	    NextTurn();
    }

    // display final scores
    DisplayScores();
}

void Game::Redo(void) {
	ASSERT(!IsClockRunning());
	ASSERT(CanRedo());

	Turn turn = *miRedo;
	miRedo++;

	ASSERT(miActiveHand->Name() == turn.HandName());

	Move move = Move(turn);	
	if (move.IsResign()) {
		miActiveHand->Resign(mStockBag);

	} else {
        // remove played/swapped tiles from the hand
        Tiles tiles = Tiles(move);
        miActiveHand->RemoveTiles(tiles);

        // draw replacement tiles from the stock bag
        Tiles draw = turn.Draw();
		miActiveHand->AddTiles(draw);
		mStockBag.RemoveTiles(draw);

	    if (move.InvolvesSwap()) {
		    // return swapped tiles to the stock bag
		    mStockBag.AddTiles(tiles);

	    } else {
	        // place played tiles on the board
            mBoard.PlayMove(move);
    
            // update the hand's score
			unsigned points = turn.Points();
			ASSERT(points == mBoard.ScoreMove(move));
	        miActiveHand->AddScore(points);
		}
	}

	//  If it was the first turn, it no longer is.
    mBestRunLength = 0;

	ActivateNextHand();

	ASSERT(!IsClockRunning());
}

unsigned Game::Redundancy(void) const {
	ASSERT(mRedundancy > 0);

	return mRedundancy;
}

void Game::Restart(void) {
	ASSERT(!IsClockRunning());

	// return all played tiles to the stock bag
	Tiles played_tiles = Tiles(mBoard);
	mStockBag.AddTiles(played_tiles);
	mBoard.MakeEmpty();

	// return all hand tiles to the stock bag
	Hands::Iterator i_hand;
	for (i_hand = mHands.begin(); i_hand != mHands.end(); i_hand++) {
		Tiles hand_tiles = Tiles(*i_hand);
		mStockBag.AddTiles(hand_tiles);
		i_hand->Restart();
	}

	// redo the initial draws
	miRedo = mHistory.begin();
	for (unsigned i_hand = 0; i_hand < mHands.Count(); i_hand++) {
		ASSERT(miRedo->Points() == 0);

		Move move = Move(*miRedo);
		ASSERT(move.IsPass());

		String hand_name = miRedo->HandName();
		miActiveHand = mHands.Find(hand_name);

		Tiles draw_tiles = miRedo->Draw();
		ASSERT(draw_tiles.Count() == mHandSize);

		miActiveHand->AddTiles(draw_tiles);
		mStockBag.RemoveTiles(draw_tiles);

		miRedo++;
	}

    // the hand with the best "run" gets the first turn
	FindBestRun();

	ASSERT(IsPaused());
	ASSERT(!CanUndo());
}

int Game::Seconds(Hand &rHand) const {
	// read the clock of a particular hand
	int result = rHand.Seconds(); 
	if (mSecondsPerHand > 0) {
		result = mSecondsPerHand - result;
		if (result < 0) {
			// TODO
		}
	}

	return result;
}

unsigned Game::SecondsPerHand(void) const {
	return mSecondsPerHand;
}

void Game::StartClock(void) {
	ASSERT(!IsOver());
	miActiveHand->StartClock();
}

void Game::StopClock(void) {
	miActiveHand->StopClock();
}

GameStyleType Game::Style(void) const {
    return mStyle;
}

void Game::Undo(void) {
	ASSERT(!IsClockRunning());
	ASSERT(CanUndo());

	miRedo--;
	Turn turn = *miRedo;

	String hand_name = turn.HandName();
	miActiveHand = mHands.Find(hand_name);

	//  Roll back the first-move info.
    mBestRunLength = turn.BestRun();

	Move move = Move(turn);
    Tiles tiles = Tiles(move);
	if (move.IsResign()) {
		miActiveHand->Unresign(mStockBag, tiles);

	} else {
        // return drawn tiles to the stock bag
        Tiles draw = turn.Draw();
		miActiveHand->RemoveTiles(draw);
		mStockBag.AddTiles(draw);

		// add played/swapped tiles back into the hand
        miActiveHand->AddTiles(tiles);

	    if (move.InvolvesSwap()) {
		    // remove swapped tiles from the stock bag
		    mStockBag.RemoveTiles(tiles);

	    } else {
	        // remove played tiles from the board
            mBoard.UnplayMove(move);
    
            // update the hand's score
			unsigned points = turn.Points();
	        miActiveHand->SubtractScore(points);
		}
	}
}


// inquiry methods

bool Game::CanRedo(void) const {
	bool result = (miRedo != mHistory.end());

	return result;
}

bool Game::CanUndo(void) const {
	unsigned turn = mHistory.Index(miRedo);
	bool result = (turn > mHands.Count());

	return result;
}

bool Game::HasEmptyCell(Cell const &rCell) const {
    bool result = mBoard.HasEmptyCell(rCell);
    
    return result; 
}

bool Game::HasUnsavedChanges(void) const {
	return mUnsavedChanges;
}

bool Game::IsClockRunning(void) const {
    bool result = miActiveHand->IsClockRunning();
     
    return result;
}

bool Game::IsLegalMove(Move const &rMove) const {
    char const *reason;
	bool result = IsLegalMove(rMove, reason);

	return result;
}

bool Game::IsLegalMove(Move const &rMove, char const *&rReason) const {
    unsigned stock = CountStock();
    bool result = true;
    
	if (!mBoard.IsValidMove(rMove, rReason)) {
	    result = false;
	} else if (mBestRunLength > 0 && !rMove.IsResign()
		&& (rMove.Count() != mBestRunLength || rMove.InvolvesSwap())) {
	    rReason = "FIRST";
        result = false;
    } else if (rMove.IsPureSwap() && rMove.Count() > stock) {
	    rReason = "STOCK";
        result = false;
    }

    return result;
}

bool Game::IsOver(void) const {
    bool result = false;
    
    // The game is over if and only if: 
	//    (1) the stock bag is empty and one of the hands has gone out 
	// or (2) all hands have resigned.

	if (IsStockEmpty() && mHands.HasAnyGoneOut()) {
        result = true;
    } else if (mHands.HaveAllResigned()) {
		result = true;
	}

    return result;
}

bool Game::IsPaused(void) const {
	bool result = !IsClockRunning() && !IsOver();

	return result;
}

bool Game::IsStockEmpty(void) const {
	bool result = (CountStock() == 0);

	return result;
}

File: trunk/Java/GoldTile/src/goldtile/Game.java
// File:     Game.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  Game class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;

public class Game
    implements ReadGame
{
    // enums, sorted by name
    public static enum Ending { RESIGNED, STUCK, WENT_OUT; };
    public static enum Tense { PAST, PRESENT; };

    // per-instance fields, sorted by type   
    private Board board = new Board();        // extensible playing surface
    final private boolean clientFlag;         // immutable
    private boolean unsavedChangeFlag = false;
    final private Hands hands;                // all hands being played
    private int mustPlay = 0;  // min number of tiles to play, 0 after the 1st turn
    final private ReadGameOpt opt;            // immutable
    private String bestRunReport = null;
    private String filespec = null;           // file for load/save
    private String firstTurnMessage = null;
    final private Tiles stockBag;
    private Turn redo = null;   // current index in the history, null means end
    final private Turns history = new Turns();
    
    // static fields
    private static Game currentInstance = null;
    
    // constructors

    public Game(GameOpt opt, HandOpt[] handOpts) {
        assert opt != null;
        opt.validate();
        assert handOpts != null;
        assert handOpts.length >= opt.getHandsDealt();

        // Replace the active game.
        currentInstance = this;

        // Save a copy of the options.
        this.opt = new GameOpt(opt);
        
        clientFlag = true; // TODO
        
        // Re-seed the psedo-random generator.
        opt.reseedGenerator();
        
        // Make a new batch of tiles.
        Tile.newGame(opt);

        // Add tiles to the stock bag.
        stockBag = new Tiles();
        if (clientFlag) {
            for (int i = 0; i < opt.getTilesPerCombo(); i++) {
                // Generate all possible tiles.
                stockBag.addAllCombos();
            }
        } else {
            // TODO
        }
        Console.printf("\nPlaced %s in the stock bag.\n",
            StringExt.plural(countStock(), "tile") );
        
        hands = new Hands(opt.getHandsDealt(), handOpts);
        assert hands.size() == opt.getHandsDealt();
        
        // The deal:  each hand pulls tiles from the stock bag.
        final int handSize = opt.getHandSize();
        int iHand = 0;
        for (Hand hand : hands) {
            assert !hand.isClockRunning();
            final Tiles dealt = pullTiles(handSize, iHand);
            assert dealt.size() == handSize : dealt;

            // Record the deal in the history.
            history.add(iHand, dealt);
            iHand++;
        }
        assert history.size() == opt.getHandsDealt();
        Console.printLine();
        
        // The hand with the best "run" gets to go first.
        findBestRun();

        // Set the redo iterator to the end of the history.
        redo = null;

        // Pretend this game has been saved, for convenience.
        unsavedChangeFlag = false;

        assert isPaused();
        assert !canUndo();
        assert !canRedo();
    }

    // methods, sorted by name
    
    public void add(Turn turn) {
        assert !getPlayable().isClockRunning();

        if (redo != null) {
           history.clearToEnd(redo);
        }
        history.addLast(turn);
        redo = null;
        unsavedChangeFlag = true;
    }
    
    @Override
    final public boolean canRedo() {
        return redo != null;
    }
    
    @Override
    final public boolean canUndo() {
        if (redo == null) {
            return history.size() > opt.getHandsDealt();
        } else {
            return false; // TODO
        }
    }

    @Override
    public UserMessage checkMove(ReadMove move) {
        assert move != null;
        
        final UserMessage result = board.checkMove(move);
        if (result != null) {
            return result;

        } else if (mustPlay > 0
                && !move.isResignation()
                && move.countTilesPlaced() != mustPlay)
        {
            // first turn but didn't resign, nor play correct number of tiles
            assert move.countTilesPlaced() < mustPlay : mustPlay;
            return UserMessage.FIRST_TURN;

        } else if (move.isPureSwap() 
                && move.size() > countStock()) {
            // swap but not enough tiles in stock
            return UserMessage.STOCK;

        } else {
            return null;
        }
    }

    public static void closeCurrentInstance() {
        currentInstance = null;
    }
    
    @Override
    public Board copyBoard() {
        return new Board(board);
    }
    
    @Override
    final public int countStock() {
        assert stockBag != null;
        
        return stockBag.size();
    }
    
    private String describeScores() {
        String result = "";
        
        for (Hand hand : hands) {
            result += hand.describeScore();
        }
        
        return result;
    }
    
    private String describeStatus() {
        final int stock = countStock();
        
        return String.format(
                "%s\n%s's turn, %s remaining in the stock bag\n\n%s\n",
                describeScores(),
                getPlayable().getName(),
                StringExt.plural(stock, "tile"),
                board.describe());      
    }

    private void findBestRun() {
        assert board != null;
        assert board.isEmpty() : board;
        assert hands != null;
        assert hands.size() > 0 : hands;
        
        final Hands.BestRun bestRun = hands.findBestRun();
        
        // Make sure the run will fit on the board.
        mustPlay = Cell.limitPlay(bestRun.length);

        firstTurnMessage = String.format(
                "%s plays first and must place %s on the (empty) board.\n",
                getPlayable().getName(), 
                StringExt.plural(mustPlay, "tile"));

        bestRunReport = bestRun.report + "\n" + firstTurnMessage;
        Console.print(bestRunReport);
        
        assert bestRunReport != null;
        assert firstTurnMessage != null;
        assert mustPlay > 0 : mustPlay;
    }
    
    private Strings findWinningHands() {
        assert hands.size() > 0;

        final int topScore = hands.findMaxScore();
        final Strings result = new Strings();
        
        for (Hand hand : hands) {
            if (hand.getScore() == topScore) {
                result.addLast(hand.getName());
            }
        }

        return result;    
    }
    
    public void finishTurn(ReadMove move) {
        assert move != null;
        assert isLegalMove(move);
        assert getPlayable().isClockRunning();

        stopClock();

        Tiles draw = new Tiles();
        int points = 0;
        
        if (move.isResignation()) {
            final Tiles tiles = hands.resign();
            stockBag.addAll(tiles);

        } else {
            // Remove played/swapped tiles from the playable hand.
            final Tiles tiles = move.copyTiles();
            hands.removeTiles(tiles);

            // Attempt to draw replacement tiles from the stock bag.
            final int tileCount = tiles.size();
            if (tileCount > 0) {
                final int iHand = hands.getPlayableIndex();
                draw = pullTiles(tileCount, iHand);
                assert !move.involvesSwap() || draw.size() == tileCount;
            }

            if (move.isPureSwap()) {
                // Return tiles to the stock bag.
                stockBag.addAll(tiles);
                Console.printf("%s returned %s to the stock bag.\n",
                        getPlayable().getName(), 
                        StringExt.plural(tileCount, "tile") );

            } else if (move.doesPlace()) {
                // Place tiles on the board.
                move.place(board);

                // Update the hand's score.
                points = board.score(move);
                hands.addScore(points);
           }
       }

       final int handIndex = hands.getPlayableIndex();
       history.add(handIndex, mustPlay, move, points, draw);

       //  If it *was* the first turn, it no longer is.
       mustPlay = 0;

       assert !getPlayable().isClockRunning();
    }
    
    private void firstTurnConsole() {
        assert GoldTile.control == Display.CONSOLE;
        assert isFirstTurn();
        assert isPaused();
        
        Console.printLine();
        if (GoldTile.enableGui) {
            //invokeLater();
        }
        startClock();
        
        final ReadHand playable = getPlayable();
        final ReadHandOpt handOpt = playable.getOpt();
        
        ReadMove move;
        if (handOpt.isAutomatic()) {
            assert clientFlag;
            assert !canRedo();
            
            move = playable.chooseMoveAutomatic();

        } else if (handOpt.isRemote()) {
            move = playable.chooseMoveRemote();
            
        } else {
            assert handOpt.isLocalUser();
            
            for (;;) {
                move = playable.chooseMoveConsole(mustPlay);
                final UserMessage reason = checkMove(move);
                if (reason == null) {
                    break;
                }
                
                Console.printf(
                        "\nThat isn't a legal move because:\n %s\n",
                        reason.message);
                Console.print("Please try again ...\n\n");
                Console.print(firstTurnMessage);
            }
        }
        
        finishTurn(move);
    }
    
    @Override
    public ReadBoard getBoard() {
        return board;    
    }
    
    @Override
    public Ending getEnding() {
        /*
         * The game is over if and only if: 
         *     (1) the stock bag is empty and a hand has gone out 
         * or (2) all hands have resigned
         * or (3) none of the last 7 moves placed any tiles on the board. 
         */
        int lpi = history.lastPlaceIndex();
        int st = opt.getStuckThreshold();
        int fir = history.findIndex(redo);
        if (isStockEmpty() && hands.hasAnyGoneOut()) {
            return Ending.WENT_OUT;
        } else if (hands.haveAllResigned()) {
            return Ending.RESIGNED;
        } else if (history.lastPlaceIndex() + opt.getStuckThreshold() <= 
                   history.findIndex(redo))
        {
            return Ending.STUCK;
        } else {
           return null;
        }
    }
    
    @Override
    public ReadHand getHand(int iHand) {
        return hands.get(iHand);
    }
    
    public static Game getInstance() {
        return currentInstance;
    }
    
    @Override
    public int getMustPlay() {
        return mustPlay;
    }
    
    @Override
    public ReadGameOpt getOpt() {
        return opt;
    }
    
    @Override
    public ReadHand getPlayable() {
        return hands.getPlayable();
    }
    
    @Override
    public int getPlayableIndex() {
        return hands.getPlayableIndex();
    }
    
    @Override
    public int getSeconds(int iHand) {
        // Read a hand's clock.
        int result = hands.get(iHand).getSecondsUsed();

        final GameStyle style = opt.getStyle();
        if (style.hasTimeLimit()) {
            // counting down
            final int timeLimit = opt.getSecondsPerHand();
            result = timeLimit - result;
        }
        
        return result;
    }
    
    public static GameStyle getStyle() {
        if (currentInstance != null) {
            final ReadGameOpt opt = currentInstance.getOpt();
            return opt.getStyle();
        } else {            
            return GameStyle.NONE;
        }
    }
    
    @Override
    public int[] getUnplayableIndices() {
        return hands.getUnplayableIndices();
    }
    
    @Override
    public Strings getUserNames() {
        final Strings result = new Strings();
        
        for (Hand hand : hands) {
            final ReadHandOpt handOpt = hand.getOpt();
            if (handOpt.isLocalUser()) {
                final String name = handOpt.getPlayerName();
                if (!result.contains(name)) {
                    result.add(name);
                }
            }
        }
        
        return result;    
    }
    
    public static boolean haveInstance() {
        return currentInstance != null;
    }
    
    @Override
    public boolean isFirstTurn() {
        return getMustPlay() > 0;
    }
    
    @Override
    public boolean isLegalMove(ReadMove move) {
        assert move != null;
        
        final UserMessage reason = checkMove(move);
        
        return reason == null;
    }
    
    @Override
    public boolean isOutOfTime() {
        if (!opt.getStyle().hasTimeLimit()) {
            return false;
        }
        
        final long have = opt.getMillisecondsPerHand();
        final long used = getPlayable().getMillisecondsUsed();

        return (used >= have);
    }
    
    @Override
    public boolean isOver() {
        return getEnding() != null;
    }

    @Override
    final public boolean isPaused() {
        return !getPlayable().isClockRunning() && !isOver();
    }

    @Override
    public boolean isPlayable(int iHand) {
        return hands.getPlayableIndex() == iHand;
    }
    
    @Override
    public boolean isStockEmpty() {
        return countStock() == 0;
    }

    public void nextHand() {
        assert !getPlayable().isClockRunning();

        // Skip over hands which have resigned.
        hands.nextWorking();
        unsavedChangeFlag = true;

        assert !getPlayable().isClockRunning();
    }
        
    private void nextTurnConsole() {
        assert GoldTile.control == Display.CONSOLE;
        assert !isFirstTurn();
        assert isPaused();

        startClock();
        Console.print(describeStatus());
        
        final ReadHand playable = getPlayable();
        final ReadHandOpt handOpt = playable.getOpt();

        ReadMove move;
        if (handOpt.isAutomatic()) {
            assert clientFlag;
            assert !canRedo();
            
            move = playable.chooseMoveAutomatic();

        } else if (handOpt.isRemote()) {
            move = playable.chooseMoveRemote();
            
        } else {
            assert handOpt.isLocalUser();
            
            for (;;) {
                move = playable.chooseMoveConsole(mustPlay);
                final UserMessage reason = checkMove(move);
                if (reason == null) {
                    break;
                }
                
                Console.printf("\nThat isn't a legal move because:\n %s\n",
                        reason.message);
                Console.print("Please try again ...\n\n");
                Console.print(describeStatus());
            }
        }
        
        finishTurn(move);
    }
    
    public void playConsole() {
        assert GoldTile.control == Display.CONSOLE;
        
        firstTurnConsole();
        
        while (!isOver()) {
            nextHand();
            nextTurnConsole();
        }
        Console.print(reportEndBonus());
        
        if (hands.size() > 1) {
            // Display final scores.
            Console.printf("%s\n", reportScores(Tense.PAST) );
        }
    }
    
    private Tiles pullTiles(int count, int handIndex) {
        assert count >= 0 : count;
        assert handIndex >= 0 : handIndex;
        assert handIndex < hands.size() : handIndex;
        
        Tiles result = null;
        
        if (clientFlag) {
            result = stockBag.pullRandom(count);
        } else {  // server
            // TODO
        }
        
        hands.addTiles(handIndex, result);

        final int actualCount = result.size();
        if (actualCount > 0) {
            Console.printf("%s drew %s from the stock bag.\n",
                    getPlayable().getName(),
                    StringExt.plural(actualCount, "tile") );
        }

        return result;
    }

    public void removeTiles(Tiles tiles) {
        assert tiles != null;
        
        hands.removeTiles(tiles);    
    }
    
    public String reportBestRun() {
        assert bestRunReport != null;
        
        return bestRunReport;    
    }
    

    public String reportEndBonus() {
        assert isOver();
        
        String result = "";
        
        final Ending ending = getEnding();
        final ReadHand playable = getPlayable();
        final String handName = playable.getName();

        if (ending == Ending.WENT_OUT) {
            // Start writing the report.
            result += String.format("%s went out with %s.\n\n",
                    handName,
                    StringExt.plural(playable.getScore(), "point") );

            // The playable hand scores a point for each tile in another hand.
            for (Hand hand : hands) {
                if (hand != playable) {  
                    final int tileCount = hand.countContents();
                    final int pointCount = tileCount; // TODO?
                
                    if (pointCount > 0) {
                        hands.addScore(pointCount);

                        result += String.format(
                                "Add %s for the %s held by %s.\n", 
                                StringExt.plural(pointCount, "point"),
                                StringExt.plural(tileCount, "tile"),
                                handName);
                    }
                }
            }
            result += "\n";
        }

        result += String.format("%s ended up with %s.\n\n",
            handName,
            StringExt.plural(playable.getScore(), "point"));

        return result;
    }
    
    private String reportScores(Tense tense) {
        assert tense != null;
        assert hands.size() > 0;
        
        String result = "";
        
        final Strings winningHands = findWinningHands();
        if (winningHands.size() == 1) {
            final String winner = winningHands.getFirst();
            result += String.format("%s %s the game.\n",
                    winner,
                    tense == Tense.PAST ? "won" : "is winning");
        } else {
            assert !winningHands.isEmpty();
            final String winners = winningHands.join(" and ");
            result += String.format("%s %s for first place.\n",
                    winners,
                    tense == Tense.PAST ? "tied" : "are tied");
        }

        return result;        
    }
    
    public void startClock() {
        assert !isOver();
        assert !getPlayable().isClockRunning();

        hands.startClock();
        assert getPlayable().isClockRunning();
    }
    
    public void stopClock() {
        assert getPlayable().isClockRunning();
        
        hands.stopClock();
    }
    
    public void togglePause() {
        assert !isOver();
        
        if (getPlayable().isClockRunning()) {
            hands.stopClock();
        } else {
            hands.startClock();
        }
    }
}
